cpp/test/sg/fil_test.cu

```diff
@@ -247,9 +248,21 @@ class BaseFilTest : public testing::TestWithParam<FilTestParams> {
     CUDA_CHECK(cudaFree(mask_d));
   }
 
+  void apply_softmax(float* class_scores) {
+    float max = *std::max_element(class_scores, &class_scores[ps.num_classes]);
+    for (int i = 0; i < ps.num_classes; ++i)
+      class_scores[i] = expf(class_scores[i] - max);
+    float sum =
+      std::accumulate(class_scores, &class_scores[ps.num_classes], 0.0f);
+    for (int i = 0; i < ps.num_classes; ++i) class_scores[i] /= sum;
+  }
+
   void transform(float f, float& proba, float& output) {
     if ((ps.output & fil::output_t::AVG) != 0) {
-      f *= (1.0f / ps.num_trees);
+      if (ps.leaf_algo == fil::leaf_algo_t::GROVE_PER_CLASS)
+        f /= ps.num_trees / ps.num_classes;
```

Comment:
How about  `*= num_trees / num_classes` ? seems like that would parallel the other branch more closely.

Also, I think braces for ifs are always nice even if not strictly necessary.

Reply:
A floating-point division would be less precise, if we have, say 100K trees and 10K classes. And an integer division in the form you suggest would yield 0.
As for braces, should we change the FIL formatting Google Doc? Also, Andy just asked to remove parentheses from this line in this review, so we should align on what to do here.

Reply:
whichever way I interpret the new expression, it is inaccurate in one or another supported case. E.g. floats cannot hold intermediates our size as precisely